theano/sandbox/cuda/acdc.py

```diff
@@ -0,0 +1,309 @@
+#
+# The following is based on acdc_wrapper.cu
+# from https://github.com/mdenil/acdc-torch
+#
+
+import os
+
+import theano
+from theano.sandbox.cuda import GpuOp
+from theano.sandbox.cuda.basic_ops import as_cuda_ndarray_variable, gpu_contiguous
+
+import cuda_ndarray
+
+class FastACDC(GpuOp):
+
+    def c_headers(self):
+        return ['<stdio.h>', '<cufft.h>']
+
+    def c_libraries(self):
+        return ['cufft']
+
+    def make_node(self, x, A, Ab, D, Db):
+        x_ = gpu_contiguous(as_cuda_ndarray_variable(x))
+        A_ = gpu_contiguous(as_cuda_ndarray_variable(A))
+        Ab_ = gpu_contiguous(as_cuda_ndarray_variable(Ab))
+        D_ = gpu_contiguous(as_cuda_ndarray_variable(D))
+        Db_ = gpu_contiguous(as_cuda_ndarray_variable(Db))
+        for t in [x_, A_, Ab_, D_, Db_]:
+            assert t.dtype == 'float32'
+        return theano.Apply(self, [x_, A_, Ab_, D_, Db_], [x_.type()])
```

Comment:
If the output shape is the same as x, I would add this method in the code:
```
def infer_shape(self, node, shape):
    return [shape[0]]
```
Otherwise, what is the shape relation of the output again the input?